pallets/file-system/src/utils.rs

```diff
@@ -445,16 +442,28 @@ where
             }
         }
 
-        // TODO: Initialise challenges properly constructing the key for this particular file.
-        let file_key = FileKeyHasher::<T>::hash(&location.encode());
-        let challenges = vec![file_key];
+        // Compute the file key hash.
+        let file_key = Self::compute_file_key(
+            file_metadata.owner.clone(),
+            location.clone(),
+            file_metadata.size,
+            file_metadata.fingerprint,
+        );
 
-        // Check that the forest proof is valid.
-        <T::ProofDealer as shp_traits::ProofsDealerInterface>::verify_forest_proof(
-            &bsp_id,
-            challenges.as_slice(),
-            &forest_proof,
-        )?;
+        // Verify the proof of non-inclusion.
+        let proven_keys: BTreeSet<MerkleHash<T>> =
+            <T::ProofDealer as shp_traits::ProofsDealerInterface>::verify_forest_proof(
+                &bsp_id,
+                &[file_key],
+                &non_inclusion_forest_proof,
+            )?;
+
+        // Ensure that the file key IS NOT part of the BSP's forest.
+        // Note: The runtime is responsible for adding and removing keys, computing the new root and updating the BSP's root.
+        ensure!(
+            !proven_keys.contains(&file_key),
+            Error::<T>::ExpectedNonInclusionProof
+        );
 
         // TODO: Generate challenges for the key proof properly.
```

Comment:
My original idea was that this could be as simple as using the `file_key` as a "seed" for generating the key challenges. After all, it has to be something that the provider knows beforehand to be able to respond to it. Although not perfect, I think it is good enough, specially for a first version at least.

Reply:
Will be adding this in another pr for simplicity.